src/fit-element.ts

```diff
@@ -0,0 +1,219 @@
+import { TemplateResult } from 'lit-html';
+import { render as shadyRender } from 'lit-html/lib/shady-render';
+
+import { ClassConstructor } from '.';
+
+declare interface Window {
+    ShadyCSS?: any;
+}
+
+/**
+ * A lit-html template construction function.
+ */
+export type TemplateFunction<P> = (props: P) => TemplateResult;
+
+/**
+ * A value constructor.
+ */
+export type Transformer = (value: string) => any;
+
+/**
+ * Attribute observer configuration.
+ *
+ * If a constructor function is present, it will be invoked when an
+ * attribute changes with the attribute's new value. If null is passed
+ * or a property is set the value will be left untouched.
+ */
+export type AttributeDescriptors<OP> = {
+    [key in keyof OP]: Transformer | null;
+};
+
+/**
+ * A 💪-html decorated class.
+ */
+export type FitDecorated<B extends ClassConstructor<HTMLElement>, OP, RP> =
+    B & FitElementConstructor<OP, RP>;
+
+/**
+ * A 💪-html decorated element.
+ */
+export interface FitElement<OP, RP> {
+    ownProps: OP;
+    renderProps: RP;
+    template: TemplateFunction<RP>;
+
+    attributeChangedCallback(name: string, oldValue: string, newValue: string);
+    connectedCallback();
+    disconnectedCallback();
+    enqueueRender();
+    render();
+}
+
+/**
+ * The constructor for a 💪-html element.
+ */
+export interface FitElementConstructor<OP, RP> {
+    observedAttributes: string[];
+    properties: AttributeDescriptors<OP>;
+
+    new(...args: any[]): FitElement<OP, RP>;
+}
+
+/* tslint:disable:max-line-length */
+
+/**
+ * Creates a subclass of the given HTML element that uses lit-html rendering and listens
+ * for attribute and property changes.
+ *
+ * @param {TemplateFunction<RP>} templ The lit-html templating function.
+ * @param {AttributeDescriptors<OP>} desc Property descriptors to enable the attribute & property listening.
+ * @returns The decorator function.
+ * @template OP, RP
+ */
+export default function withFit<OP, RP = OP>(templ: TemplateFunction<RP>, desc?: AttributeDescriptors<OP>) {
+    return <T extends ClassConstructor<HTMLElement>>(base: T): FitDecorated<T, OP, RP> => {
+        const Element = class extends base {
```

Comment:
We probably want to check against inheritance from `HTMLElement` here, because we are using things like `attachShadow` and `nodeName`. I can build this.

Reply:
Runtime checks? I'd rather not do this because we have TS static ones.

Reply:
People may still use this library without typescript...